### PR TITLE
Connect disconnect update

### PR DESCRIPTION
**Pull request by zakorgy**
Depends on https://github.com/szeged/servo/pull/86#issuecomment-266022043.
Update for the GATTServer's connect() disconnect() methods corresponding to the specification. Added the `clean-up-disconnected-device` and `garbage-collection-the-connection` methods and functions invalidating cached IDs in the middle layer.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes

### CODE REVIEW

File: components/script/dom/bluetoothdevice.rs [hunk 1]
```diff
@@ -133,6 +150,41 @@ impl BluetoothDevice {
...
+                                  .collect::<Vec<()>>();
+
+        let mut descriptor_map = self.attribute_instance_map.2.borrow_mut();
+
+        // Step 4, 7.
+        let _ = descriptor_map.drain().map(|(_, d)| d.get().set_represented_descriptor_to_null()).collect::<Vec<()>>();
+
+        // Step 8.
+        self.upcast::<EventTarget>().fire_bubbling_event(atom!("gattserverdisconnected"));
+
```
Comment (dati91) [inline components/script/dom/bluetoothdevice.rs]: remove empty line

File: tests/wpt/mozilla/tests/mozilla/bluetooth/getCharacteristics/get-different-characteristics-after-reconnection-with-uuid.html [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        .then(() => gattServer.connect())
+        .then(() => gattServer.getPrimaryService(heart_rate.name))
+        .then(service => service.getCharacteristics(body_sensor_location.name))
+        .then(characteristics2 => [characteristics1, characteristics2])
+    })
+    .then(characteristics_arrays => {
+        for (let i = 1; i < characteristics_arrays.length; i++) {
+            assert_equals(characteristics_arrays[0].length, characteristics_arrays[i].length);
+        }
+
```
Comment (dati91) [inline tests/wpt/mozilla/tests/mozilla/bluetooth/getCharacteristics/get-different-characteristics-after-reconnection-with-uuid.html]: remove empty line

File: tests/wpt/mozilla/tests/mozilla/bluetooth/getCharacteristics/get-different-characteristics-after-reconnection.html [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        .then(() => gattServer.connect())
+        .then(() => gattServer.getPrimaryService(heart_rate.name))
+        .then(service => service.getCharacteristics())
+        .then(characteristics2 => [characteristics1, characteristics2])
+    })
+    .then(characteristics_arrays => {
+        for (let i = 1; i < characteristics_arrays.length; i++) {
+            assert_equals(characteristics_arrays[0].length, characteristics_arrays[i].length);
+        }
+
```
Comment (dati91) [inline tests/wpt/mozilla/tests/mozilla/bluetooth/getCharacteristics/get-different-characteristics-after-reconnection.html]: remove empty line

File: tests/wpt/mozilla/tests/mozilla/bluetooth/getDescriptors/get-different-descriptors-after-reconnection-with-uuid.html [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        .then(() => gattServer.getPrimaryService(generic_access.name))
+        .then(service => service.getCharacteristic(device_name.name))
+        .then(characteristic => characteristic.getDescriptors(number_of_digitals.name))
+        .then(descriptors2 => [descriptors1, descriptors2])
+    })
+    .then(descriptors_arrays => {
+        for (let i = 1; i < descriptors_arrays.length; i++) {
+            assert_equals(descriptors_arrays[0].length, descriptors_arrays[i].length);
+        }
+
```
Comment (dati91) [inline tests/wpt/mozilla/tests/mozilla/bluetooth/getDescriptors/get-different-descriptors-after-reconnection-with-uuid.html]: remove empty line

File: tests/wpt/mozilla/tests/mozilla/bluetooth/getDescriptors/get-different-descriptors-after-reconnection.html [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        .then(() => gattServer.getPrimaryService(generic_access.name))
+        .then(service => service.getCharacteristic(device_name.name))
+        .then(characteristic => characteristic.getDescriptors(number_of_digitals.name))
+        .then(descriptors2 => [descriptors1, descriptors2])
+    })
+    .then(descriptors_arrays => {
+        for (let i = 1; i < descriptors_arrays.length; i++) {
+            assert_equals(descriptors_arrays[0].length, descriptors_arrays[i].length);
+        }
+
```
Comment (dati91) [inline tests/wpt/mozilla/tests/mozilla/bluetooth/getDescriptors/get-different-descriptors-after-reconnection.html]: remove empty line

File: tests/wpt/mozilla/tests/mozilla/bluetooth/getPrimaryServices/get-different-services-after-reconnection-with-uuid.html [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        .then(() => gattServer.disconnect())
+        .then(() => gattServer.connect())
+        .then(() => gattServer.getPrimaryServices(heart_rate.name))
+        .then(services2 => [services1, services2])
+    })
+    .then(services_arrays => {
+        for (let i = 1; i < services_arrays.length; i++) {
+            assert_equals(services_arrays[0].length, services_arrays[i].length);
+        }
+
```
Comment (dati91) [inline tests/wpt/mozilla/tests/mozilla/bluetooth/getPrimaryServices/get-different-services-after-reconnection-with-uuid.html]: remove empty line

File: tests/wpt/mozilla/tests/mozilla/bluetooth/getPrimaryServices/get-different-services-after-reconnection.html [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        .then(() => gattServer.disconnect())
+        .then(() => gattServer.connect())
+        .then(() => gattServer.getPrimaryServices())
+        .then(services2 => [services1, services2])
+    })
+    .then(services_arrays => {
+        for (let i = 1; i < services_arrays.length; i++) {
+            assert_equals(services_arrays[0].length, services_arrays[i].length);
+        }
+
```
Comment (dati91) [inline tests/wpt/mozilla/tests/mozilla/bluetooth/getPrimaryServices/get-different-services-after-reconnection.html]: remove empty line

File: components/script/dom/bluetoothdevice.rs [hunk 8]
```diff
@@ -133,6 +150,41 @@ impl BluetoothDevice {
...
+        // TODO: Step 2: Implement activeAlgorithms internal slot for BluetoothRemoteGATTServer.
+
+        // Step 3: We don't need `context`, we get the attributeInstanceMap from the device.
+        // https://github.com/WebBluetoothCG/web-bluetooth/issues/330
+
+        let mut service_map = self.attribute_instance_map.0.borrow_mut();
+
+        // Step 4, 5.
+        let _ = service_map.drain().map(|(_, s)| s.get().set_represented_service_to_null()).collect::<Vec<()>>();
```
Comment (dati91) [inline components/script/dom/bluetoothdevice.rs]: If we don't have to check if it succeeds, why not send 3 vec to the lower level like `SetRepresentedToNull(Vec<dev>, Vec<char>, Vec<desc>, sender)` then we can remove the `GATTLevel Enum`. (I don't see where you use the the Device enum)

Comment (zakorgy) [reply components/script/dom/bluetoothdevice.rs]: I agree that `GATTLevel Enum` should be removed from the current implementation, but when we start implementing the [Service change events](https://webbluetoothcg.github.io/web-bluetooth/#service-change-events) in `Step 6` we must invalidates only services/characterisitcs/descriptors that have been changed. `GATTLevel::Device` can be removed until [Revoke bluetooth acces](https://webbluetoothcg.github.io/web-bluetooth/#revoke-bluetooth-access) Step 4. is missing from our implementation.
Or I just remove the enum and we will deal with this when we need it.

Comment (dati91) [reply components/script/dom/bluetoothdevice.rs]: I think we should deal with the device stuff later.

File: components/script/dom/bluetoothremotegattserver.rs [hunk 9]
```diff
@@ -47,6 +47,42 @@ impl BluetoothRemoteGATTServer {
...
+        self.get_bluetooth_thread().send(
+            BluetoothRequest::GATTServerDisconnect(String::from(self.Device().Id()), sender)).unwrap();
+        let server = receiver.recv().unwrap();
+        match server {
+            Ok(_) => {
+                Ok(())
+            },
+            Err(error) => {
+                Err(Error::from(error))
+            },
```
Comment (mmatyas) [inline components/script/dom/bluetoothremotegattserver.rs]: Put the match branches to one line

File: components/script/dom/bluetoothdevice.rs [hunk 10]
```diff
@@ -133,6 +150,37 @@ impl BluetoothDevice {
         descriptor_map.insert(descriptor.instance_id.clone(), MutJS::new(&bt_descriptor));
         return bt_descriptor;
     }
+
```
Comment (dati91) [inline components/script/dom/bluetoothdevice.rs]: remove empty line

File: tests/wpt/mozilla/tests/mozilla/bluetooth/getCharacteristics/disconnect-invalidates-objects-with-uuid.html [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    window.testRunner.setBluetoothMockDataSet(adapter_type.heart_rate);
+    return window.navigator.bluetooth.requestDevice({
+        filters: [{services: [heart_rate.name]}]
+    })
+    .then(device => device.gatt.connect())
+    .then(gattServer => {
+        return gattServer.getPrimaryService(heart_rate.name)
+        .then(service => service.getCharacteristics(body_sensor_location.name))
+        .then(c => {
+            let characteristics = [].concat(c);
```
Comment (dati91) [inline tests/wpt/mozilla/tests/mozilla/bluetooth/getCharacteristics/disconnect-invalidates-objects-with-uuid.html]: why do you need this concat?

Comment (zakorgy) [reply tests/wpt/mozilla/tests/mozilla/bluetooth/getCharacteristics/disconnect-invalidates-objects-with-uuid.html]: This is for ensure that the `characteristics` variable is an iterable object (in this case an Array), by default the `c` return value is not iterable for some reason. Without this, we can't use the [let for ... of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of) syntax in [line27](https://github.com/szeged/servo/pull/89/files/8f0cf985b3a8b6dfd93a09013a04d8a1a7865a10..fac70aca4a206a326a52918e0c02048854ab7ef0#diff-884d56cd987682247c62a5262484dad3R27), which needs characteristics to be iterable.

File: components/script/dom/bluetoothremotegattserver.rs (hunk 9 above)
Comment (zakorgy) [reply components/script/dom/bluetoothremotegattserver.rs]: We can replace this match and the line above with `receiver.recv().unwrap().map_err(Error::from)`.

**Comment by dati91** [comment]
lgtm
you can squash it

**Comment by dati91** [comment]
rebase it with the latest servo master and it can merge

File: tests/wpt/mozilla/tests/mozilla/bluetooth/getCharacteristics/disconnect-invalidates-objects-with-uuid.html (hunk 11 above)
Comment (zakorgy) [reply tests/wpt/mozilla/tests/mozilla/bluetooth/getCharacteristics/disconnect-invalidates-objects-with-uuid.html]: Forget it, we can get rid of these.

**Pull request merged**